File: jsNCAA/scripts/_sim_spreads.mjs
import fs from "fs";
const h = JSON.parse(fs.readFileSync("data/history.json", "utf8"));

const picks = [];
for (const r of h.runs || []) {
  if (r.burnIn) continue;
  for (const g of r.games || []) {
    if (!g.sPick || g.sPick === "PASS") continue;
    if (!g.sResult || g.sResult === "PUSH") continue;
    picks.push({
      result: g.sResult === "WIN" ? 1 : 0,
      pCover: g.pCover || 0,
      sDiff: Math.abs(g.sDiff || 0),
      line: Math.abs(g.line || 0),
      conf: g.sConf || "low",
      isFav: (g.sPick || "").includes("-"),
    });
  }
}

console.log(`Total graded spread picks: ${picks.length}\n`);

function test(label, filterFn) {
  const filtered = picks.filter(filterFn);
  const w = filtered.filter(p => p.result).length;
  const l = filtered.length - w;
  const pct = filtered.length > 0 ? (w / filtered.length * 100).toFixed(1) : "n/a";
  const units = (w - l * 1.1).toFixed(1);
  const roi = filtered.length > 0 ? (parseFloat(units) / filtered.length * 100).toFixed(1) : "n/a";
  console.log(`${label.padEnd(50)} ${String(w + "-" + l).padEnd(10)} ${pct}%  ${units}u  ROI: ${roi}%`);
}

console.log("═══ Current (baseline) ═══");
test("ALL picks", p => true);

console.log("\n═══ By pCover ═══");
test("pCover >= 0.58", p => p.pCover >= 0.58);
test("pCover >= 0.60", p => p.pCover >= 0.60);
test("pCover >= 0.62", p => p.pCover >= 0.62);
test("pCover >= 0.63", p => p.pCover >= 0.63);
test("pCover >= 0.65", p => p.pCover >= 0.65);
test("pCover >= 0.67", p => p.pCover >= 0.67);
test("pCover >= 0.70", p => p.pCover >= 0.70);

console.log("\n═══ By sDiff ═══");
test("sDiff >= 2", p => p.sDiff >= 2);
test("sDiff >= 3", p => p.sDiff >= 3);
test("sDiff >= 4", p => p.sDiff >= 4);
test("sDiff >= 5", p => p.sDiff >= 5);
test("sDiff >= 6", p => p.sDiff >= 6);

console.log("\n═══ By spread size ═══");
test("line <= 5", p => p.line <= 5);
test("line <= 8", p => p.line <= 8);
test("line <= 10", p => p.line <= 10);
test("line <= 12", p => p.line <= 12);
test("line <= 15", p => p.line <= 15);
test("line 3-10", p => p.line >= 3 && p.line <= 10);
test("line 3-12", p => p.line >= 3 && p.line <= 12);

console.log("\n═══ Fav vs Dog ═══");
test("Favorites only", p => p.isFav);
test("Underdogs only", p => !p.isFav);
test("Dogs pCover>=0.63", p => !p.isFav && p.pCover >= 0.63);
test("Dogs pCover>=0.65", p => !p.isFav && p.pCover >= 0.65);
test("Favs pCover>=0.65", p => p.isFav && p.pCover >= 0.65);

console.log("\n═══ Combined filters ═══");
test("pCover>=0.63 AND sDiff>=3", p => p.pCover >= 0.63 && p.sDiff >= 3);
test("pCover>=0.63 AND line<=10", p => p.pCover >= 0.63 && p.line <= 10);
test("pCover>=0.63 AND line<=12", p => p.pCover >= 0.63 && p.line <= 12);
test("pCover>=0.65 AND sDiff>=3", p => p.pCover >= 0.65 && p.sDiff >= 3);
test("pCover>=0.65 AND line<=10", p => p.pCover >= 0.65 && p.line <= 10);
test("pCover>=0.65 AND line<=12", p => p.pCover >= 0.65 && p.line <= 12);
test("pCover>=0.65 AND sDiff>=4", p => p.pCover >= 0.65 && p.sDiff >= 4);
test("pCover>=0.67 AND line<=10", p => p.pCover >= 0.67 && p.line <= 10);
test("pCover>=0.67 AND line<=12", p => p.pCover >= 0.67 && p.line <= 12);
test("pCover>=0.65 AND line 3-12", p => p.pCover >= 0.65 && p.line >= 3 && p.line <= 12);
test("pCover>=0.63 AND line 3-12", p => p.pCover >= 0.63 && p.line >= 3 && p.line <= 12);

console.log("\n═══ Dogs + combined ═══");
test("Dogs pCover>=0.63 line<=10", p => !p.isFav && p.pCover >= 0.63 && p.line <= 10);
test("Dogs pCover>=0.63 line<=12", p => !p.isFav && p.pCover >= 0.63 && p.line <= 12);
test("Dogs pCover>=0.65 line<=12", p => !p.isFav && p.pCover >= 0.65 && p.line <= 12);
test("Dogs pCover>=0.65 line<=10", p => !p.isFav && p.pCover >= 0.65 && p.line <= 10);
test("Dogs pCover>=0.67 line<=12", p => !p.isFav && p.pCover >= 0.67 && p.line <= 12);

File: pyNCAA/scripts/_sim_spreads.py
#!/usr/bin/env python3
# scripts/_sim_spreads.py — Spread simulation sweep
import json

h = json.loads(open("data/history.json").read())
picks = []
for r in (h.get("runs") or []):
    if r.get("burnIn"): continue
    for g in (r.get("games") or []):
        if not g.get("sPick") or g["sPick"] == "PASS": continue
        if g.get("sResult") not in ("WIN", "LOSS"): continue
        picks.append({
            "result": 1 if g["sResult"] == "WIN" else 0,
            "pCover": g.get("pCover", 0), "sDiff": abs(g.get("sDiff", 0)),
            "line": abs(g.get("line", 0)), "conf": g.get("sConf", "low"),
            "isFav": "-" in (g.get("sPick") or ""),
        })

print(f"Total graded spread picks: {len(picks)}\n")

def test(label, filter_fn):
    filtered = [p for p in picks if filter_fn(p)]
    w = sum(p["result"] for p in filtered)
    l = len(filtered) - w
    pct = f"{w / len(filtered) * 100:.1f}" if filtered else "n/a"
    units = f"{(w - l * 1.1):.1f}"
    roi = f"{float(units) / len(filtered) * 100:.1f}" if filtered else "n/a"
    print(f"{label:<50} {w}-{l!s:<10} {pct}%  {units}u  ROI: {roi}%")

print("=== Current (baseline) ===")
test("ALL picks", lambda p: True)
print("\n=== By pCover ===")
for t in [0.58, 0.60, 0.62, 0.63, 0.65, 0.67, 0.70]:
    test(f"pCover >= {t}", lambda p, t=t: p["pCover"] >= t)
print("\n=== By sDiff ===")
for t in [2, 3, 4, 5, 6]:
    test(f"sDiff >= {t}", lambda p, t=t: p["sDiff"] >= t)
print("\n=== By spread size ===")
for lo, hi in [(0,5),(0,8),(0,10),(0,12),(0,15),(3,10),(3,12)]:
    test(f"line {lo}-{hi}" if lo else f"line <= {hi}", lambda p, lo=lo, hi=hi: lo <= p["line"] <= hi)
print("\n=== Fav vs Dog ===")
test("Favorites only", lambda p: p["isFav"])
test("Underdogs only", lambda p: not p["isFav"])
test("Dogs pCover>=0.63", lambda p: not p["isFav"] and p["pCover"] >= 0.63)
test("Dogs pCover>=0.65", lambda p: not p["isFav"] and p["pCover"] >= 0.65)
test("Favs pCover>=0.65", lambda p: p["isFav"] and p["pCover"] >= 0.65)
print("\n=== Combined filters ===")
test("pCover>=0.63 AND sDiff>=3", lambda p: p["pCover"] >= 0.63 and p["sDiff"] >= 3)
test("pCover>=0.63 AND line<=10", lambda p: p["pCover"] >= 0.63 and p["line"] <= 10)
test("pCover>=0.63 AND line<=12", lambda p: p["pCover"] >= 0.63 and p["line"] <= 12)
test("pCover>=0.65 AND sDiff>=3", lambda p: p["pCover"] >= 0.65 and p["sDiff"] >= 3)
test("pCover>=0.65 AND line<=10", lambda p: p["pCover"] >= 0.65 and p["line"] <= 10)
test("pCover>=0.65 AND line<=12", lambda p: p["pCover"] >= 0.65 and p["line"] <= 12)
test("pCover>=0.65 AND sDiff>=4", lambda p: p["pCover"] >= 0.65 and p["sDiff"] >= 4)
print("\n=== Dogs + combined ===")
test("Dogs pCover>=0.63 line<=10", lambda p: not p["isFav"] and p["pCover"] >= 0.63 and p["line"] <= 10)
test("Dogs pCover>=0.63 line<=12", lambda p: not p["isFav"] and p["pCover"] >= 0.63 and p["line"] <= 12)
test("Dogs pCover>=0.65 line<=12", lambda p: not p["isFav"] and p["pCover"] >= 0.65 and p["line"] <= 12)
test("Dogs pCover>=0.65 line<=10", lambda p: not p["isFav"] and p["pCover"] >= 0.65 and p["line"] <= 10)
test("Dogs pCover>=0.67 line<=12", lambda p: not p["isFav"] and p["pCover"] >= 0.67 and p["line"] <= 12)
